server/comms/rpc/rpctypes.go

```diff
@@ -0,0 +1,358 @@
+	Sig     string `json:"sig"`
+}
+
+// Error is returned as part of a response to indicate that an error occurred
+// during method execution.
+type Error struct {
+	Code    int    `json:"code"`
+	Message string `json:"message"`
+}
+
+// NewError is a constructor for an Error.
+func NewError(code int, msg string) *Error {
+	return &Error{
+		Code:    code,
+		Message: msg,
+	}
+}
+
+// ResponsePayload is the payload for a Response-type Message.
+type ResponsePayload struct {
+	// Result is the payload, if successful, else nil.
+	Result json.RawMessage `json:"result,omitempty"`
+	// Error is the error, or nil if none was encountered.
+	Error *Error `json:"error,omitempty"`
+}
+
+// MessageType indicates the type of message. MessageType is typically the first
+// switch checked when examining a message, and how the rest of the message is
+// decoded depends on its MessageType.
+type MessageType uint8
+
+const (
+	InvalidMessageType MessageType = iota // 0
+	Request                               // 1
+	Response                              // 2
+	Notification                          // 3
+)
+
+// Message is the primary messaging type for websocket communications.
+type Message struct {
+	// Type is the message type.
+	Type MessageType `json:"type"`
+	// Route is used for requests and notifications, and specifies a handler for
+	// the message.
+	Route string `json:"route,omitempty"`
+	// ID is a unique number that is used to link a response to a request.
+	ID uint64 `json:"id",omitempty`
+	// Payload is any data attached to the message. How Payload is decoded
+	// depends on the Route.
+	Payload json.RawMessage `json:"payload",omitempty`
+}
+
+// DecodeMessage decodes a *Message from JSON-formatted bytes.
+func DecodeMessage(b []byte) (*Message, error) {
+	msg := new(Message)
+	err := json.Unmarshal(b, &msg)
+	if err != nil {
+		return nil, err
+	}
+	return msg, nil
+}
+
+// NewRequest is the constructor for a Request-type *Message.
+func NewRequest(id uint64, route string, payload interface{}) (*Message, error) {
+	encoded, err := json.Marshal(payload)
+	if err != nil {
+		return nil, err
+	}
+	return &Message{
+		Type:    Request,
+		Payload: json.RawMessage(encoded),
+		Route:   route,
+		ID:      id,
+	}, nil
+}
+
+// NewResponse encodes the result and creates a Response-type *Message.
+func NewResponse(id uint64, result interface{}, rpcErr *Error) (*Message, error) {
+	encResult, err := json.Marshal(result)
+	if err != nil {
+		return nil, err
+	}
+	resp := &ResponsePayload{
+		Result: encResult,
+		Error:  rpcErr,
+	}
+	encResp, err := json.Marshal(resp)
+	if err != nil {
+		return nil, err
+	}
+	return &Message{
+		Type:    Response,
+		Payload: json.RawMessage(encResp),
+		ID:      id,
+	}, nil
+}
+
+// Response attempts to decode the payload to a *ResponsePayload. Response will
+// return an error if the Type is not Response.
+func (msg *Message) Response() (*ResponsePayload, error) {
+	if msg.Type != Response {
```

Comment:
`Response` is really just a convenience method for parsing the common response structure from the message payload. I'm not sure if notifications will share a common payload structure too, so a method like this may or may not be needed. The proposed message protocol leaves the notification payload type unspecified, so right now it can be of any type. We may find that notifications would also benefit from a common payload structure. I was trying not to speculate too much.